### PR TITLE
Adding unrecognized method in BaseExtensionRestHandler

### PR DESCRIPTION
**Pull request by vibrantvarun**
### Description
Adding unrecognized method in BaseExtensionRestHandler for Extensions Architecture. The method calls unrecognized method from BaseRestHandler in OpenSearch.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by joshpalis — DISMISSED** [review]

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java [hunk 1]
```diff
@@ -127,4 +135,58 @@ protected ExtensionRestResponse createJsonResponse(RestRequest request, RestStat
...
+     * @param detail
+     * @return an String that contains the message.
+     */
+    public static final String unrecognized(RestRequest request, Set<String> invalids, Set<String> candidates, String detail) {
+        StringBuilder message = new StringBuilder(
+            String.format(Locale.ROOT, "request [%s] contains unrecognized %s%s: ", request.path(), detail, invalids.size() > 1 ? "s" : "")
+        );
+        boolean first = true;
+
+        for (Iterator var7 = invalids.iterator(); var7.hasNext(); first = false) {
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:154]: Should be 
```suggestion
        for (final String invalid : invalids) {
```

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java [hunk 2]
```diff
@@ -127,4 +135,58 @@ protected ExtensionRestResponse createJsonResponse(RestRequest request, RestStat
...
+                    message.append("[").append((String) keys.get(0)).append("]");
+                } else {
+                    message.append("any of ").append(keys.toString());
+                }
+
+                message.append("?");
+            }
+        }
+
+        return message.toString();
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:190]: I don't thin so this method is copy/pasted correctly [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/BaseRestHandler.java#L131)

**Review by owaiskazi19 — CHANGES_REQUESTED** [review]
Added comments

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java (hunk 1 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:154]: The variable first is always initialized as false in every iteration. If I accept you commit suggestion then I have to add another line which will initialized first as false for every iteration.  Therefore my suggestion would be to not change anything.

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java (hunk 2 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:190]: ACK

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java (hunk 1 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:154]: See the next comment. Copy/paste the method exactly. That will override this change

Comment (vibrantvarun) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java]: Got it. Pushed the updated changes.
Thanks

**Comment by owaiskazi19** [comment]
@dbwiddis are you fine with this change?

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java [hunk 3]
```diff
@@ -127,4 +134,58 @@ protected ExtensionRestResponse createJsonResponse(RestRequest request, RestStat
             return new ExtensionRestResponse(request, status, fieldName + ": " + responseStr);
         }
     }
+
+    /**
+     * Returns a String message of the detail of any unrecognized error occurred
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:139]: Suggestion: add a sentence indicating it is intended for use in error messages to be returned to a user.

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java [hunk 4]
```diff
@@ -127,4 +134,58 @@ protected ExtensionRestResponse createJsonResponse(RestRequest request, RestStat
...
+     *
+     * @param request The request that caused the exception
+     * @param invalids invalid strings
+     * @param candidates candidates involved
+     * @param detail
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:144]: This needs a description of what it actually is.  Looking at line 154 it appears to be "a noun describing the category or type of invalid strings to be passed".    Also apparently it is supposed to be singular.   TLDR, I couldn't tell from reading this javadoc what I am supposed to do with this method.

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java [hunk 5]
```diff
@@ -127,4 +134,58 @@ protected ExtensionRestResponse createJsonResponse(RestRequest request, RestStat
             return new ExtensionRestResponse(request, status, fieldName + ": " + responseStr);
         }
     }
+
+    /**
+     * Returns a String message of the detail of any unrecognized error occurred
+     *
+     * @param request The request that caused the exception
+     * @param invalids invalid strings
+     * @param candidates candidates involved
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:143]: Are candidates "involved"?  I don't know what this means. Looking at the code it appears to be using levenshtein distance to provide a "did you mean" functionality.  So say so here.  "A set of words that are most likely to be the valid strings determined invalid, to be suggested to the user."

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java [hunk 6]
```diff
@@ -127,4 +134,58 @@ protected ExtensionRestResponse createJsonResponse(RestRequest request, RestStat
             return new ExtensionRestResponse(request, status, fieldName + ": " + responseStr);
         }
     }
+
+    /**
+     * Returns a String message of the detail of any unrecognized error occurred
+     *
+     * @param request The request that caused the exception
+     * @param invalids invalid strings
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java:142]: "Strings from the request which were unable to be understood."

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Ugh.   So if we keep this, we need much better javadoc.  See specifics below.

But I am really not fond of duplicating code.  Is there any possible way of creating a thin wrapper method that delegates to the original?

```java
    // add this method to both BaseRestHandler and BaseExtensionRestHandler
    /**
     * Insert javadocs here
     */
    protected final String unrecognized(
        final RestRequest request,
        final Set<String> invalids,
        final Set<String> candidates,
        final String detail
    ) {
        return unrecognizedStrings(request, invalids, candidates, detail);
    }

    // change signature of existing method in BaseRestHandler
    /**
     * Insert javadocs here
     */
    public static String unrecognizedStrings(
        final RestRequest request,
        final Set<String> invalids,
        final Set<String> candidates,
        final String detail
    ) {
        // current unrecognized() code
    }
```

Or, this works in `BaseExtensionRestHandler` without changing the signature in `BaseRestHandler`.  Ugly and I wouldn't do it, but it's possible. Inheritance is wierd.
```java
    /**
     * Insert javadocs here
     */
    protected final String unrecognized(
        final RestRequest request,
        final Set<String> invalids,
        final Set<String> candidates,
        final String detail
    ) {        
        return new BaseRestHandler() {
            @Override
            public String getName() {
                return null;
            }

            @Override
            protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
                return null;
            }            
            
            public String unrecognizedStrings(
                final RestRequest request,
                final Set<String> invalids,
                final Set<String> candidates,
                final String detail
                ) {
                return super.unrecognized(request, invalids, candidates, detail);                
            }
        }.unrecognizedStrings(request, invalids, candidates, detail);
    }
```

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java (hunk 4 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java]: Updated

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java (hunk 5 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java]: Updated

File: src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java (hunk 6 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java]: updated

**Review by dbwiddis — APPROVED** [review]
LGTM! Thanks for the refactor!

**Review by owaiskazi19 — APPROVED** [review]
Much cleaner!

**Pull request merged**